core/src/main/scala/kafka/server/KafkaServer.scala

```diff
@@ -283,38 +305,44 @@ class KafkaServer(val config: KafkaConfig, time: Time = SystemTime) extends Logg
           val controllerId = ZkUtils.getController(zkClient)
           ZkUtils.getBrokerInfo(zkClient, controllerId) match {
             case Some(broker) =>
-              if (channel == null || prevController == null || !prevController.equals(broker)) {
-                // if this is the first attempt or if the controller has changed, create a channel to the most recent
-                // controller
-                if (channel != null) {
-                  channel.disconnect()
-                }
-                channel = new BlockingChannel(broker.getBrokerEndPoint(config.interBrokerSecurityProtocol).host,
-                  broker.getBrokerEndPoint(config.interBrokerSecurityProtocol).port,
-                  BlockingChannel.UseDefaultBufferSize,
-                  BlockingChannel.UseDefaultBufferSize,
-                  config.controllerSocketTimeoutMs)
-                channel.connect()
+              // if this is the first attempt, if the controller has changed or if an exception was thrown in a previous
+              // attempt, connect to the most recent controller
+              if (ioException || broker != prevController) {
+
+                ioException = false
+
+                if (prevController != null)
+                  selector.close(connectionId(prevController))
+
+                val brokerEndPoint = broker.getBrokerEndPoint(config.interBrokerSecurityProtocol)
+                selector.connect(
+                  connectionId(broker),
+                  new InetSocketAddress(brokerEndPoint.host, brokerEndPoint.port),
+                  Selectable.USE_DEFAULT_BUFFER_SIZE,
+                  Selectable.USE_DEFAULT_BUFFER_SIZE
+                )
```

Comment:
OK, I see. I'll check if turning off metadata in `NetworkClient` can be done without too much complexity.